Platform/OpenNtfsDxe/Index.c

```diff
@@ -0,0 +1,918 @@
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+extern UINTN  mFileRecordSize;
+extern UINTN  mIndexRecordSize;
+STATIC UINT64 mBufferSize;
+
+STATIC
+VOID
+FreeNode (
+  IN NTFS_FILE  *Node    OPTIONAL,
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  if (Node != Context->RootNode) {
+    if (Node != NULL) {
+      FreePool (Node);
+    }
+  }
+}
+
+STATIC
+VOID
+PopElement (
+  IN FSHELP_CTX *Context
+  )
+{
+  STACK_ELEMENT *Element;
+
+  ASSERT (Context != NULL);
+
+  Element = Context->CurrentNode;
+  Context->CurrentNode = Element->Parent;
+  FreeNode (Element->Node, Context);
+  FreePool (Element);
+}
+
+STATIC
+VOID
+FreeStack (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  while (Context->CurrentNode != NULL) {
+    PopElement (Context);
+  }
+}
+
+STATIC
+VOID
+GoUpALevel (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  if (Context->CurrentNode->Parent == NULL) {
+    return;
+  }
+  PopElement (Context);
+}
+
+STATIC
+EFI_STATUS
+PushNode (
+  IN FSHELP_CTX       *Context,
+  IN NTFS_FILE        *Node,
+  IN FSHELP_FILETYPE  FileType
+  )
+{
+  STACK_ELEMENT *Next;
+
+  ASSERT (Context != NULL);
+  ASSERT (Node != NULL);
+
+  Next = AllocateZeroPool (sizeof (STACK_ELEMENT));
+  if (Next == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Next->Node = Node;
+  Next->Type = FileType & ~FSHELP_CASE_INSENSITIVE;
+  Next->Parent = Context->CurrentNode;
+  Context->CurrentNode = Next;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+GoToRoot (
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  FreeStack (Context);
+  return PushNode (Context, Context->RootNode, FSHELP_DIR);
+}
+
+STATIC
+EFI_STATUS
+FindFileIter (
+  IN CHAR16           *Name,
+  IN FSHELP_FILETYPE  FileType,
+  IN NTFS_FILE        *Node,
+  IN FSHELP_ITER_CTX  *Context
+  )
+{
+  INTN Result;
+
+  ASSERT (Name != NULL);
+  ASSERT (Node != NULL);
+  ASSERT (Context != NULL);
+
+  if (FileType == FSHELP_UNKNOWN) {
+    return EFI_NOT_FOUND;
+  }
+
+  if ((FileType & FSHELP_CASE_INSENSITIVE) != 0) {
+    Result = OcStriCmp (Context->Name, Name);
+  } else {
+    Result = StrCmp (Context->Name, Name);
+  }
+
+  if (Result != 0) {
+    return EFI_NOT_FOUND;
+  }
+
+  *Context->FoundNode = Node;
+  *Context->FoundType = FileType;
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+NtfsDirHook (
+  IN CHAR16           *Name,
+  IN FSHELP_FILETYPE  FileType,
+  IN NTFS_FILE        *Node,
+  IN EFI_FILE_INFO    *Info
+  )
+{
+  EFI_STATUS  Status;
+  INT64       *Index = (INT64 *) &Info->FileSize;
+  EFI_TIME    Time = { 1970, 01, 01, 00, 00, 00, 0, 0, 0, 0, 0};
+  INT32       Mtime;
+
+  ASSERT (Name != NULL);
+  ASSERT (Node != NULL);
+  ASSERT (Info != NULL);
+
+  if ((Name[0] == L'.') &&
+     ((Name[1] == 0) || ((Name[1] == L'.') &&
+     (Name[2] == 0)))) {
+    FreeFile (Node);
+    return EFI_NOT_FOUND;
+  }
+
+  if ((*Index)-- != 0) {
```

Comment:
Can you comment this? I have no idea what this means honestly

Reply:
This is called alternative intelligence or spaghetti code. FileReadDir() is called to get the next entry of the opened earlier directory. EFI_NTFS_FILE.DirIndex is used as a counter of already read entries. The main cycle, which looks through directory's contents stored on disk, is located in ListFile(). It would be too easy, from the former developers' point of view, to check, whether loop counter equals DirIndex, inside this function. That's why, NtfsDirHook() is called every iteration, it decrements Index and when Index equals zero, then new entry is found and information about it can be returned. Otherwise EFI_NOT_FOUND is returned and ListFile() goes on the next iteration.

Reply:
And this intentionally clobbers EFI_FILE_INFO.FileSize for directions (to be the current index), yes? The correct value, i.e. the size of the directory contents, is not set anywhere, correct? I agree bothering to calculate the directory size during listing is stupid, but imo it should at least be commented it is omitted (this is not optional by the spec as far as I can tell). Is the following where the size is set to 0, so that the caller always gets this value? https://github.com/acidanthera/OpenCorePkg/pull/332/files/f7c51ec0ec6b3bd77a99f38de9c5bb7b0af5b261#diff-138d91b89f447bda352d32ddb2fb00d256725785c953800c588e9a440d80b88bR193

Reply:
This intentionally repurposes `FileSize` for the time being, the correct value is set later https://github.com/acidanthera/OpenCorePkg/pull/332/files/f7c51ec0ec6b3bd77a99f38de9c5bb7b0af5b261#diff-138d91b89f447bda352d32ddb2fb00d256725785c953800c588e9a440d80b88bR212
According to UEFI Spec 2.8 you don't need to return the size of the directory contents. 
`Size`: Size of the EFI_FILE_INFO structure, including the Null-terminated FileName string.
`FileSize`: The size of the file in bytes.
`PhysicalSize`: The amount of physical space the file consumes on the file system volume.

Reply:
Can’t check much right now as I’m on phone, but this was the snippet I looked at yesterday:
“On directories, the file size is determined by the contents of the directory and cannot be
changed by setting FileSize. On directories, FileSize is ignored during a SetInfo()."
I don’t think it’s optional per-spec. Please somewhat comment the code to make clear the value is butchered and we break the spec intentionally.

Reply:
Also maybe add a FIXME to refactor this some day. It’s opaque, fugly, and breaks Strict Aliasing.